### ISSUE TITLE
Question about custom function parameters

### ISSUE THREAD

**Issue by pebezo** [lunet-io/scriban#19]
Hi,

First of all, super cool project! We have a current project that uses DotLiquid. While nice, the fact that they don't have a proper parser is holding them back. I'm looking into switching to Sciban and was wondering if I could do something like this:

`{{ myfunction foo bar }}`

In this example, `myfunction` would be a function that I would define in the C# project (similar to drops and tags in DotLiquid) while `foo` and `bar` would be values that I would receive in the C# function. In the C# code I would actually need to see values `foo` and `bar` (not treat them as variables). Is something like this allowed or would I have to string-ify them? Like:

`{{ myfunction "foo" "bar" }}`

Thanks.

**Comment by xoofx** [comment]
In your scenario, `foo` `bar` are expected to be variables or parameter-less functions, and if they are not defined, they will be replaced by an empty/null when calling `myfunction`. So indeed, in this case, we use string arguments.

**Comment by pebezo** [comment]
In the scenario that they're replaced with null, since they're treated as functions, is there a way to find (reflect?) the original name that was used? In other words, is there a "missing function" hook that I could use? Thanks.

**Comment by xoofx** [comment]
Currently, there is no hook, but I could easily add one. Though one thing is that it would just be a hook out of context (you would not know that it is called by `myfunction`). Let me know and I can push a small API addition for this.

**Comment by xoofx** [comment]
I have added a support for late binding for variable and member access, you have an example [here](https://github.com/lunet-io/scriban/blob/dc159942e700c59a8b80f8ffd0654e376fa5119f/src/Scriban.Tests/TestRuntime.cs#L15)
Let me know if it fits your needs and I will push a new version

**Comment by pebezo** [comment]
Wow, you're fast...  The first example with `dynamic {{ myvar }}` would not work for our use case since I only want these to be "situational", i.e. only as a parameter to certain functions. I wouldn't want to process all cases. Some people may find it useful though.

The second example, however, that looks like `dynamic {{ a.myvar }}` would work. This would allow us to have a special predefined "name" object so that `{{ myfunction name.foo }}` would get translated into `{{ myfunction "foo" }}`. 

I haven't tested yet, but I'm assuming the `target == "a"` in your test case. This way our C# function could just expect a string and the designers could use either version: `{{ myfunc "foo" }}` or `{{ myfunc name.foo }}`.

Thanks.

**Comment by xoofx** [comment]
If it is in the context of some functions, then you can use a custom `IScriptCustomFunction`. You can have a look at the [include function](https://github.com/lunet-io/scriban/blob/0706c4c504cb0705f55eb2e44154cfd24cf199bb/src/Scriban/Runtime/IncludeFunction.cs) for example.

While the argument `ScriptArray parameters` gives you access to the expanded values (not what you want), you can still have access to the AST tree through the argument `ScriptNode callerContext`. In the case of a function call, the `callerContext` would be a [`ScriptFunctionCall`](https://github.com/lunet-io/scriban/blob/0706c4c504cb0705f55eb2e44154cfd24cf199bb/src/Scriban/Runtime/ScriptFunctionCall.cs#L20) and as you can see you dig into the AST directly.

The custom script function can be then assigned to a variable (as it is for the [`include` function registration](https://github.com/lunet-io/scriban/blob/7cee9161ba337b727f4291069d9eb11cc6c2c36f/src/Scriban/Runtime/BuiltinFunctions.cs#L10) on a ScriptObject instance.

**Comment by pebezo** [comment]
Being able to access the AST via callerContext is super powerful! My use case with DotLiquid was to call "assets" that have "names" that allow numbers, letters, and dashes. 

The issue is with the dashes. Since something like  `{% myfunction foo-bar %}` was allowed with DotLiquid I was thinking to allow the same syntax with Scriban `{{ myfunction foo-bar }}`. Since Scriban evaluates the `foo-bar` expression (as it should) this won't work.

However, this is just a minor difference; I'd rather have proper parsing instead. We've decided to just ask them to wrap the 'names' in quotes.

We were able to convert all the functionality from DotLiquid (with one exception; will open a separate issue). I must say that working with the library so far has be flawless. I would try something out and it'd just work. Nice work!

Unless there's something unfinished/buggy that I'm not aware of, I would recommend that you bump the version to 1.x. This would give the impression of a more mature library (it looks pretty polished to me).

Also, it would be nice if you could find time to finish the Runtime documentation. Even a basic version (not as thorough as the Language doc) would be helpful.

**Comment by xoofx** [comment]
Thanks for the feedback, glad to hear that you were able to port from DotLiquid.

> Also, it would be nice if you could find time to finish the Runtime documentation. Even a basic version (not as thorough as the Language doc) would be helpful.

Yes, it is a long standing issue. Lots part of the code are also not commented, the public API should receive a massive update...

> Unless there's something unfinished/buggy that I'm not aware of, I would recommend that you bump the version to 1.x. This would give the impression of a more mature library (it looks pretty polished to me).

The reason for not bumping to 1.x is that for this library (and a couple of recent ones I have been authoring), I'm trying to keep a semantic versioning, so the `0.x` allows me to slightly break things. It is easy to do because the library hasn't many users and I haven't taken the time to promote it, so I know that it will have a minimal impact. 

That being said, as part of the documentation work, I have to take the time to finish polishing it (code + documentation), write a blog post about it... Once this is done, more users should be able to come, more usecases should bring updates to the API... and once it has been a bit more stress-tests (like you did), I will be more confidently bump to 1.x
